drivers/hcsr04/hcsr04.c

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (C) 2021 AUTH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_hcsr04
+ * @{
+ *
+ * @file
+ * @brief       Device driver implementation for the HC-SR04
+ *
+ * @author      Evripidis Chondromatidis <eurichon1996@gmail.com>
+ *
+ * @}
+ */
+
+#include "hcsr04.h"
+#include "hcsr04_constants.h"
+#include "hcsr04_params.h"
+
+#include "periph/gpio.h"
+#include "xtimer.h"
+
+#include <stdint.h>
+#include "assert.h"
+#include <errno.h>
+#include <math.h>
+
+#include <stdio.h>
+
+void hcsr04_int_callback(void *arg)
+{
+    uint32_t now = xtimer_now_usec();
+    hcsr04_t *p_dev = (hcsr04_t *)arg;
+
+    if (gpio_read(p_dev->params.echo_pin)) {
+        p_dev->pre_meas_t_us = now;
+    } else {
+        uint32_t dt_us_one_way = (now - p_dev->pre_meas_t_us) / 2;
+        p_dev->distance = p_dev->sound_speed * dt_us_one_way / 1000;
+    }
+
+    mutex_unlock(&p_dev->lock);
+}
+
+int hcsr04_init(hcsr04_t *dev, const hcsr04_params_t *params)
+{
+    assert(dev && params);
+
+    dev->params = *params;
+
+    mutex_init(&dev->lock);
+    dev->distance = 0;
+    dev->was_triggered = 0;
+    dev->pre_trig_t_us = xtimer_now_usec();
+    dev->pre_meas_t_us = xtimer_now_usec();
+    hcsr04_set_temp(dev, params->temperature);
+
+    if (gpio_init(dev->params.trigger_pin, GPIO_OUT)) {
+        return -EIO;
+    }
+
+    if (gpio_init_int(dev->params.echo_pin, GPIO_IN, GPIO_BOTH, hcsr04_int_callback, dev)) {
+        return -EIO;
+    }
+
+    // necessary in order for the next trigger not to fail due to time constraint 
+    xtimer_msleep(30);
+
+    return 0;
+}
+
+uint32_t hcsr04_get_max_response_time(hcsr04_t *dev) {
+    return 2000 * HCSR04_MAX_DISTANCE_MM / dev->sound_speed;
+}
+
+int hcsr04_trigger(hcsr04_t *dev)
+{
+    uint32_t now = xtimer_now_usec();
+    uint16_t min_response_time = hcsr04_get_max_response_time(dev);
+    
+    if ((now - dev->pre_trig_t_us) < min_response_time) {
+        return -EFREQ;
+    }
+    
+    dev->was_triggered = 1;
+    dev->pre_trig_t_us = now;
+    gpio_set(dev->params.trigger_pin);
+    xtimer_usleep(TRIGGER_TIME);
+    gpio_clear(dev->params.trigger_pin);
+    
+    mutex_unlock(&dev->lock);
+    mutex_lock(&dev->lock);
+
+    return 0;
+}
```

Comment:
I think it could be made private (e.g.: a static function) and call it from the `hcsr04_get_distance` function as you do, thus leaving only one function for getting the distance (by triggering the read), allows to remove the `was_triggered` field saving some bytes (and anyway if the read was not triggered an error is returned so, why not 😉 ). Same with the `hcsr04_read` function (make it private).